lib/multiplexer/multiplexer.go

```diff
@@ -0,0 +1,259 @@
+		waitContext: waitContext,
+		waitCancel:  waitCancel,
+	}, nil
+}
+
+// Mux supports having both SSH and TLS on the same listener socket
+type Mux struct {
+	sync.RWMutex
+	*log.Entry
+	Config
+	listenerClosed bool
+	sshListener    *Listener
+	tlsListener    *Listener
+	context        context.Context
+	cancel         context.CancelFunc
+	waitContext    context.Context
+	waitCancel     context.CancelFunc
+}
+
+// SSH returns listener that receives SSH connections
+func (m *Mux) SSH() net.Listener {
+	return m.sshListener
+}
+
+// TLS returns listener that receives TLS connections
+func (m *Mux) TLS() net.Listener {
+	return m.tlsListener
+}
+
+func (m *Mux) isClosed() bool {
+	m.RLock()
+	defer m.RUnlock()
+	return m.listenerClosed
+}
+
+func (m *Mux) closeListener() {
+	m.Lock()
+	defer m.Unlock()
+	// propagate close signal to other listeners
+	m.cancel()
+	if m.Listener == nil {
+		return
+	}
+	if m.listenerClosed {
+		return
+	}
+	m.listenerClosed = true
+	m.Listener.Close()
+}
+
+// Close closes listener
+func (m *Mux) Close() error {
+	m.closeListener()
+	return nil
+}
+
+// Wait waits until listener shuts down and stops accepting new connections
+// this is to workaround issue https://github.com/golang/go/issues/10527
+// in tests
+func (m *Mux) Wait() {
+	<-m.waitContext.Done()
+}
+
+// Serve is a blocking function that serves on the listening socket
+// and accepts requests. Every request is served in a separate goroutine
+func (m *Mux) Serve() error {
+	defer m.waitCancel()
+	backoffTimer := time.NewTicker(5 * time.Second)
+	defer backoffTimer.Stop()
+	for {
+		conn, err := m.Listener.Accept()
+		if err != nil {
+			if m.isClosed() {
+				return trace.ConnectionProblem(nil, "listener is closed")
+			}
+			select {
+			case <-backoffTimer.C:
+				m.Debugf("backoff on network error")
+			case <-m.context.Done():
+				return trace.ConnectionProblem(nil, "context is closing")
+			}
+		} else {
+			go m.detectAndForward(conn)
+		}
+
+	}
+}
+
+func (m *Mux) detectAndForward(conn net.Conn) {
+	conn.SetReadDeadline(m.Clock.Now().Add(m.ReadDeadline))
+	connWrapper, err := detect(conn)
+	if err != nil {
+		m.Warning(err.Error())
+		conn.Close()
+		return
+	}
+
+	conn.SetReadDeadline(time.Time{})
+	switch connWrapper.protocol {
+	case ProtoTLS:
+		select {
+		case m.tlsListener.connC <- connWrapper:
+		case <-m.context.Done():
+			connWrapper.Close()
+			return
+		}
+	case ProtoSSH:
+		if m.WithoutSSH {
+			m.Warningf("detected SSH connection, SSH is turned off")
+			connWrapper.Close()
+			return
+		}
+		select {
+		case m.sshListener.connC <- connWrapper:
+		case <-m.context.Done():
+			connWrapper.Close()
+			return
+		}
+	default:
+		// should not get there, but let's still handle this just in case
+		connWrapper.Close()
+		m.Errorf("detected but unsupported protocol: %v", connWrapper.protocol)
+	}
+}
+
+func detect(conn net.Conn) (*Conn, error) {
+	reader := bufio.NewReader(conn)
+
+	var proxyLine *ProxyLine
+	for i := 0; i < 2; i++ {
```

Comment:
I don't really see why you need a loop here:
```go
func detect(conn net.Conn) (*Conn, error) {
	reader := bufio.NewReader(conn)

	proto, err := readProto(reader)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	var proxyLine *ProxyLine
	if proto == ProtoProxy {
		proxyLine, err = ReadProxyLine(reader)
		if err != nil {
			return nil, trace.Wrap(err)
		}
		proto, err = readProto(reader)
	}
	switch proto {
	case ProtoTLS, ProtoSSH:
		return &Conn{
			protocol:  proto,
			Conn:      conn,
			reader:    reader,
			proxyLine: proxyLine,
		}, nil
	}

	return nil, trace.BadParameter("failed to detect protocol")
}

func readProto(r io.Reader) (int, error) {
	bytes, err := r.Peek(3)
	if err != nil {
		return ProtoUnknown, trace.Wrap(err, "failed to peek connection")
	}

	proto, err := detectProto(bytes)
	if err != nil {
		return ProtoUnknown, trace.Wrap(err)
	}
	return proto, nil
}
```

Reply:
I don't think your version is more readable than mine tbh, so I will keep it as is

Reply:
`2` looks like a magic number, can you add a comment at the top explaining why you loop over this twice.

Also, this code will have problems if you have more than a single layer of proxy protocol.